File: app/src/main/java/com/lcarino/bucketlist/mvp/MvpActivity.java
package com.lcarino.bucketlist.mvp;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

/**
 * @author Luis Carino.
 */

public abstract class MvpActivity<V extends MvpView, P extends MvpPresenter<V>> extends AppCompatActivity {

    protected P presenter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = createPresenter();
    }

    /**
     * Instantiate a getListPresenter instance
     *
     * @return The {@link MvpPresenter} for this view
     */
    public  abstract P createPresenter();
}

File: app/src/main/java/com/lcarino/bucketlist/mvp/kotlin/MvpActivity.kt
package com.lcarino.bucketlist.mvp.kotlin

import android.os.Bundle
import android.support.v7.app.AppCompatActivity

/**
 * Base Android MVP Activity.
 * Created by luiscarino on 3/8/17.
 */
abstract class MvpActivity<V : View, P : MvpPresenter<V>> : AppCompatActivity(), MvpPresenter<V> {

    protected var presenter: P? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        presenter = createPresenter()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onDetachView() {
        presenter?.onDetachView()
    }

    abstract fun createPresenter(): P
}
